test/tests/filter.js

```diff
@@ -0,0 +1,833 @@
+      let author = signatures[0];
+      let committer = signatures[1];
+
+      return repo.createCommit(
+        "HEAD",
+        author,
+        committer,
+        commitMessage,
+        treeOid,
+        [parent]);
+    });
+  }
+
+  beforeEach(function() {
+    var test = this;
+
+    return Repository.open(reposPath)
+      .then(function(repository) {
+        test.repository = repository;
+        return Repository.open(emptyRepoPath);
+      })
+      .then(function(emptyRepo) {
+        test.emptyRepo = emptyRepo;
+        return fse.writeFile(
+          path.join(reposPath, ".gitattributes"), 
+          "*.md filter="+ filterName +" -text", {
+          encoding: "utf-8",
+        });
+      });
+  });
+
+  afterEach(function() {
+    return Registry.unregister(filterName)
+      .catch(function(error) {
+        switch(error) {
+          case -1:
+            throw new Error("Cannot unregister filter");
+          // case -3:
+          //   throw new Error('Cannot find filter to unregister');
+          default:
+            return;
+        }
+      });
+  });
+
+  describe("Register", function() {
+    var secondFilter = "hellofilter";
+
+    after(function(done) {
+      return Registry.unregister(secondFilter)
+        .then(function() {
+          done();
+        });
+    });
+
+    it("can register a filter", function() {
+      return Registry.register(filterName, mockFilter, 0)
+        .then(function(result) {
+          assert.strictEqual(result, 0);
+        });
+    });
+
+    it("can register multiple filters", function() {
+      return Registry.register(filterName, mockFilter, 0)
+        .then(function(result) {
+          assert.strictEqual(result, 0);
+          return Registry.register(secondFilter, mockFilter, 1);
+        })
+        .then(function(result) {
+          assert.strictEqual(result, 0);
+        });
+    });
+
+    it("cannot register the same filter twice", function() {
+      return Registry.register(filterName, mockFilter, 0)
+        .then(function(result) {
+          assert.strictEqual(result, 0);
+          return Registry.register(filterName, mockFilter, 0);
+        })
+        .catch(function(error) {
+          assert.strictEqual(error.errno, -4);
+        });
+    });
+  });
+
+  describe("Unregister", function() {
+    beforeEach(function() {
+      return Registry.register(filterName, mockFilter, 0);
+    });
+
+    it("can unregister the filter", function() {
+      return Registry.unregister(filterName)
+        .then(function(result) {
+          assert.strictEqual(result, 0);
+        });
+    });
+
+    it("cannot unregister the filter twice", function() {
+      return Registry.unregister(filterName)
+        .then(function(result) { 
+          assert.strictEqual(result, 0);
+          return Registry.unregister(filterName);
+        })
+        .then(function(result) {
+          assert.fail(result, -3, "Should not have unregistered successfully");
```

Comment:
if caught, result will never `=== -3` (it's actually the errno), so this line is somewhat nonsensical.
```javascript
assert.fail("Should not have unregsitered successfully");
```
should be sufficient.